File: go/day94.go
package main

import (
	"fmt"
	"os"
	"strings"
	"bytes"
	"bufio"
)


func setCsvToBlank() {
	dbFile := "__db.csv"

	// Open file with truncate flag to clear contents
	fileHandle, err := os.OpenFile(dbFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
	if err != nil {
		panic(err)
	}
	defer fileHandle.Close()

	// Nothing to write, file is now blank
}

func addToCsv(rowValues []string) int {
	dbFile := "__db.csv"

	// Open or create file in append mode
	fileHandle, err := os.OpenFile(dbFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		panic(err)
	}
	defer fileHandle.Close()

	// map-style replacement
	for i, value := range rowValues {
		rowValues[i] = `"` + strings.ReplaceAll(value, `"`, `\"`) + `"`
	}


	// Join row values into CSV format
	row := strings.Join(rowValues, ",") + "\n"

	// Write row
	if _, err := fileHandle.Write([]byte(row)); err != nil {
		panic(err)
	}

	// Read entire file to count lines
	rows, err := os.ReadFile(dbFile)
	if err != nil {
		panic(err)
	}

	// Count number of newline characters = number of rows
	return bytes.Count(rows, []byte("\n"))
}


func trimRowValues(values []string) []string {
	unquoted := values;
	for i,value := range values {
		unquoted[i] = strings.Trim(
			strings.TrimSpace(value), 
			`"`,
		) // removes leading & trailing "
	}
	return unquoted;
}

func readCsvRow(rowId int) []string {
	dbFile := "__db.csv"
	// Read entire file to count lines
	rowBytes, err := os.ReadFile(dbFile)
	if err != nil {
		panic(err)
	}
	rows := strings.Split(string(rowBytes), "\n")
	return trimRowValues(strings.Split(rows[rowId],","));
}

func getCsvRowCount() int {
	dbFile := "__db.csv"
	// Read entire file to count lines
	rowBytes, err := os.ReadFile(dbFile)
	if err != nil {
		panic(err)
	}
	rows := strings.Split(string(rowBytes), "\n")
	return len(rows);
}

func readAllCsvRow() [][]string {

	numRows := getCsvRowCount()

	rows := make([][]string, numRows);
	for x := 0; x < numRows; x++ {
		rows[x] = readCsvRow(x);
	}
	return rows;
}

func getCsvColumnNumber(columnName string) int {
	columns := readCsvRow(0);
	for i,column := range columns {
		column = strings.TrimSpace(column);
		if column == columnName {
			return i+1;
		}
	}
	panic("Column not found");
}

func updateCsvRow( rowId int, columnName string, value string) {
	rows := readAllCsvRow();

	columnNumber := getCsvColumnNumber(columnName);

	rows[rowId][columnNumber-1] = value;

	setCsvToBlank();
	// map-style replacement
	for _, row := range rows {
		addToCsv(row);
	}
}

func findFirstCsvRow(columnName string, value string) *int {
	rows := readAllCsvRow()

	columnNumber := getCsvColumnNumber(columnName)
	columnIndex := columnNumber - 1

	for rowId := 1; rowId < len(rows); rowId++ {
		row := rows[rowId]
		if row[columnIndex] == value {
			return &rowId
		}
	}
	return nil
}


func main() {

	fmt.Println(
		"input the email address:",
	);

	reader := bufio.NewReader(os.Stdin)
	email1, _ := reader.ReadString('\n')
	email1 = strings.TrimSpace(email1)

	rowId := findFirstCsvRow("Email", email1);

	if rowId != nil {
		panic("Email is duplicate(error)");
	}

	fmt.Println(
		"input name:",
	);


	name1, _ := reader.ReadString('\n')
	name1 = strings.TrimSpace(name1)

	addToCsv([]string{email1, name1});

	fmt.Println("Email/Name successfully added to csv")
	
}

File: cSharp/day94.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Collections.Generic;

const string DbFile = "__db.csv";

static int AddToCsv(string[] rowValues)
{
    // Replace inner quotes and wrap each value in quotes
    for (int i = 0; i < rowValues.Length; i++)
    {
        rowValues[i] = "\"" + rowValues[i].Replace("\"", "\\\"") + "\"";
    }
    // Join row values into CSV format
    string row = string.Join(",", rowValues) + "\n";

    // Read entire file and count number of newline characters
    string contents = File.ReadAllText(DbFile);
    char contentLastCharacter = contents[^1];

    if (contentLastCharacter != '\n') {
        row = '\n' + row;
    }

    // Append row to file (create if not exists)
    File.AppendAllText(DbFile, row, Encoding.UTF8);

    return contents.Count(c => c == '\n');
}

static string[] TrimQuotes(string[] values)
{
    return values.Select(v => v.Trim('"')).ToArray();
}

static string[] ReadCsvRow(int rowId)
{
    string[] rows = File.ReadAllLines(DbFile);
    if (rowId < 0 || rowId >= rows.Length)
        throw new IndexOutOfRangeException("Row ID out of range");

    return TrimQuotes(rows[rowId].Split(","));
}

static int GetCsvRowCount()
{
    return File.ReadAllLines(DbFile).Length;
}

static List<string[]> ReadAllCsvRow()
{
    int numRows = GetCsvRowCount();;

    var rows = new List<string[]>(numRows);
    for (int x = 0; x < numRows; x++)
    {
        rows.Add(ReadCsvRow(x));
    }
    return rows;
}

static int GetCsvColumnNumber(string columnName)
{
    string[] columns = ReadCsvRow(0);
    for (int i = 0; i < columns.Length; i++)
    {
        if (columns[i].Trim() == columnName)
        {
            return i + 1; // 1-based index like Go version
        }
    }
    throw new Exception("Column not found");
}

static string GetCSVCellValue(int rowId, int columnNumber) {
    string[] row = ReadCsvRow(rowId);


    return row[columnNumber - 1];
}

static int[] SearchCsvRows(string columnName, string value) {
    List<string[]> rows = ReadAllCsvRow();

    int columnNumber = GetCsvColumnNumber(columnName);
    int columnIndex = columnNumber - 1;

    Stack<int> result = new Stack<int>();

    for (int x = 1; x < rows.Count; x++)
    {
        string[] row = rows[x];
        if (row[columnIndex] == value)
        {
            result.Push(x);
        }
    }
    return result.ToArray();
}

static int? FindFirstCsvRow(string columnName, string value)
{
    List<string[]> rows = ReadAllCsvRow();

    int columnNumber = GetCsvColumnNumber(columnName);
    int columnIndex = columnNumber - 1;

    Stack<int> result = new Stack<int>();

    for (int rowId = 1; rowId < rows.Count; rowId++)
    {
        string[] row = rows[rowId];
        if (row[columnIndex] == value)
        {
            return rowId;
        }
    }
    return null;
}

Console.WriteLine("Please input your email");

string email1 = Console.ReadLine();

int? rowId = FindFirstCsvRow("Email", email1);

if (rowId != null)
{
    Console.WriteLine("Error: duplicate email");
}
else
{
    Console.WriteLine("Please input name");

    string name1 = Console.ReadLine();
    
    AddToCsv([email1, name1]);
}